File: Gravical.Poker.Core/Pocket.cs
namespace Gravical.Poker.Core
{
    /// <summary>
    /// Pocket holds two distinct poker cards for a player.
    /// </summary>
    public class Pocket
    {
        public Card CardA { get; }
        public Card CardB { get; }

        public Pocket(Deck deck)
        {
            Guards.ArgumentNotNull(deck, nameof(deck));

            CardA = deck.DrawCard();
            CardB = deck.DrawCard();
        }

        public Pocket(byte[] binary)
        {
            Guards.ArgumentHasExactSize(binary, 2, nameof(binary));
            CardA = new Card(binary[0]);
            CardB = new Card(binary[1]);
        }

        public override string ToString()
        {
            return $"{CardA} {CardB}";
        }

        public Card[] ToArray()
        {
            return new[] { CardA, CardB };
        }

        public bool IsTrumped(Pocket by)
        {
            Guards.ArgumentNotNull(by, nameof(by));

            return CardA.Equals(by.CardA) || CardA.Equals(by.CardB) || CardB.Equals(by.CardA) || CardB.Equals(by.CardB);
        }

        public bool IsTrumped(TableFinal by)
        {
            Guards.ArgumentNotNull(by, nameof(by));

            return by.Contains(CardA) || by.Contains(CardB);
        }

        public byte[] ToBinary()
        {
            return new byte[] { CardA.Key, CardB.Key };
        }
    }
}

File: JavaScript/src/Pocket.mjs
import { Card } from './Card.mjs';
import { Deck } from './Deck.mjs';
import { TableFinal } from './TableFinal.mjs';

// Pocket holds two distinct poker cards for a player.
export class Pocket {

	constructor(init) {

		if (init && init instanceof Deck) {
			this.a = init.drawCard();
			this.b = init.drawCard();
		} else if (init && init instanceof Uint8Array) {
			if (init.length != 2) throw "Initializer must be two bytes";
			this.a = Card.parse(init[0]);
			this.b = Card.parse(init[1]);
		} else if (!init) {
			this.a = null;
			this.b = null;
		} else {
			throw "Initializer is of an unknown type";
		}
	}

	toString() {
		return `${this.a} ${this.b}`;
	}

	toArray() {
		return [ this.a, this.b ];
	}

	isTrumped(by) {
		if (!by) {
			throw "By is null";
		} else if (by instanceof Pocket) {
			return this.a.equals(by.a) || this.a.equals(by.b) || this.b.equals(by.a) || this.b.equals(by.b);
		} else if (by instanceof TableFinal) {
			return by.contains(this.a) || by.contains(this.b);
		} else {
			throw "Unknown by";
		}
	}

	toBinary() {
		return new Uint8Array([ this.a ? this.a.uint8() : 0, this.b ? this.b.uint8() : 0]);
	}
}
